src/checks.ml

```diff
@@ -0,0 +1,203 @@
+      if Int64.compare io_offset encoded_sizeL < 0 then (
+        if not (Int64.equal io_offset 0L) then
+          Fmt.failwith
+            "Non-integer number of entries in file: { offset = %Ld; entry_size \
+             = %d }"
+            io_offset Entry.encoded_size)
+      else
+        let first_entry = IO.read_entry io 0L in
+        let previous = ref first_entry in
+        Format.eprintf "\n%!";
+        Progress_unix.(
+          counter ~total:io_offset ~mode:`UTF8
+            ~message:"Scanning store for faults" ~pp:Progress.Units.bytes
+            ~sampling_interval:100_000 ()
+          |> with_reporters)
+        @@ fun report ->
+        io
+        |> IO.iter ~min:encoded_sizeL (fun ~off e ->
+               report encoded_sizeL;
+               if !previous.key_hash > e.key_hash then
+                 Log.err (fun f ->
+                     f "Found non-monotonic region:@,%a@,"
+                       (print_window_around off io context)
+                       ());
+               previous := e)
+  end
+
+  module Cli = struct
+    open Cmdliner
+
+    let path =
+      let open Arg in
+      required
+      @@ pos 0 (some string) None
+      @@ info ~doc:"Path to the Index store on disk" ~docv:"PATH" []
+
+    let quick_stats = Term.(const (fun root () -> Quick_stats.v ~root) $ path)
+
+    let integrity_check =
+      Term.(const (fun root () -> Integrity_check.v ~root) $ path)
+
+    let setup_log =
+      let init style_renderer level =
+        let format_reporter =
+          let report _src level ~over k msgf =
+            let k _ =
+              over ();
+              k ()
+            in
+            msgf @@ fun ?header:_ ?tags:_ fmt ->
+            match level with
+            | Logs.App ->
+                Fmt.kpf k Fmt.stderr
+                  ("@[<v 0>%a" ^^ fmt ^^ "@]@.")
+                  Fmt.(styled `Bold (styled (`Fg `Cyan) string))
+                  ">> "
+            | _ -> Fmt.kpf k Fmt.stdout ("@[<v 0>" ^^ fmt ^^ "@]@.")
+          in
+          { Logs.report }
+        in
+        Fmt_tty.setup_std_outputs ?style_renderer ();
+        Logs.set_level level;
+        Logs.set_reporter format_reporter
+      in
+      Term.(const init $ Fmt_cli.style_renderer () $ Logs_cli.level ())
+
+    let main () : empty =
+      let default =
+        let default_info =
+          let doc = "Check and repair Index data-stores." in
+          Term.info ~doc "index-fsck"
+        in
+        Term.(ret (const (`Help (`Auto, None))), default_info)
+      in
+      Term.(
+        eval_choice default
+          [
+            ( quick_stats $ setup_log,
+              Term.info ~doc:"Print high-level statistics about the store."
+                "stat" );
+            ( integrity_check $ setup_log,
+              Term.info
+                ~doc:"Search the store for integrity faults and corruption."
+                "check" );
+          ]
+        |> (exit : unit result -> _));
+      assert false
```

Comment:
Why do you need this assert false? Also, can you use `unit` instead of `empty` here ?

Reply:
The `assert false` and `empty` are both used to make it clear at the type-level that this function never returns. Returning `unit` would also work, but doesn't make that explicit.

Reply:
The `assert false` and `empty` are both used to make it clear at the type-level that this function never returns. Returning `unit` would also work, but doesn't make that explicit.

Reply:
I see, thanks!